File: app/src/main/java/com/chess/cryptobot/content/balance/BalanceHolder.java
package com.chess.cryptobot.content.balance;

import android.content.Context;

import androidx.fragment.app.Fragment;

import com.chess.cryptobot.content.ContextHolder;
import com.chess.cryptobot.content.Preferences;
import com.chess.cryptobot.exceptions.ItemNotFoundException;
import com.chess.cryptobot.market.Market;
import com.chess.cryptobot.market.MarketFactory;
import com.chess.cryptobot.model.Balance;
import com.chess.cryptobot.model.ViewItem;
import com.chess.cryptobot.task.BalanceUpdateTask;
import com.chess.cryptobot.task.CoinImageTask;
import com.chess.cryptobot.view.MainFragment;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;

public class BalanceHolder extends ContextHolder {
    private boolean hasKeys;
    private Map<String, Boolean> bittrexStatuses;
    private Map<String, Boolean> livecoinStatuses;

    public BalanceHolder(Fragment fragment) {
        super(fragment);
        bittrexStatuses = new HashMap<>();
        livecoinStatuses = new HashMap<>();
        checkIfHasKeys();
    }

    private void checkIfHasKeys() {
        hasKeys = true;
        MarketFactory factory = new MarketFactory();
        for (Market market: factory.getMarkets(this)) {
            if (market.keysIsEmpty()) {
                hasKeys = false;
                break;
            }
        }
    }

    @Override
    public Preferences initPrefs(Context context) {
        return new BalancePreferences(context);
    }

    public void initViewItems(Set<String> coinNames) {
        coinNames.forEach(coinName -> addItemToList(new Balance(coinName)));
    }

    public void add(String coinName) {
        Balance balance = new Balance(coinName);
        if (!getViewItems().contains(balance))
            add(new Balance(coinName));
    }

    @Override
    public void add(ViewItem viewItem) {
        super.add(viewItem);
        Balance balance = (Balance) viewItem;
        balance.setStatuses(livecoinStatuses.get(balance.getName()), bittrexStatuses.get(balance.getName()));
        updateImage(balance);
        if (hasKeys) updateAmount(balance);
        getMainActivity().updateBot();
    }

    @Override
    public synchronized void remove(ViewItem item) {
        super.remove(item);
        getMainActivity().updateBot();
    }

    @Override
    public void updateItem(ViewItem item) {
        Balance balance = (Balance) item;
        balance.setStatuses(livecoinStatuses.get(balance.getName()), bittrexStatuses.get(balance.getName()));
        updateImage(balance);
        if (hasKeys) updateAmount(balance);
    }

    private void updateImage(Balance balance) {
        MainFragment fragment = getMainFragment();
        if (fragment != null) {
            CoinImageTask task = new CoinImageTask(this);
            task.execute(balance);
        }
    }

    private void updateAmount(Balance balance) {
        BalanceUpdateTask task = new BalanceUpdateTask(this);
        task.execute(balance);
    }

    public Balance getBalanceByPosition(int position) throws ItemNotFoundException {
        MainFragment balanceActivity = getMainFragment();
        if (balanceActivity != null) {
            String coinName = balanceActivity.itemNameByPosition(position);
            return (Balance) this.getItemByName(coinName);
        }
        return null;
    }

    public void setMinBalance(String coinName, Double minBalance) {
        BalancePreferences preferences = (BalancePreferences) getPrefs();
        preferences.setMinBalance(coinName,minBalance);
    }

    public void setCurrencyStatus(Map<String, Boolean> bittrexStatuses, Map<String, Boolean> livecoinStatuses) {
        this.bittrexStatuses = bittrexStatuses;
        this.livecoinStatuses = livecoinStatuses;
    }
}

File: app/src/main/java/com/chess/cryptobot/content/balance/BalanceHolder.kt
package com.chess.cryptobot.content.balance

import android.content.Context
import androidx.fragment.app.Fragment
import com.chess.cryptobot.content.ContextHolder
import com.chess.cryptobot.content.Preferences
import com.chess.cryptobot.exceptions.ItemNotFoundException
import com.chess.cryptobot.market.MarketFactory
import com.chess.cryptobot.model.Balance
import com.chess.cryptobot.model.ViewItem
import com.chess.cryptobot.task.BalanceUpdateTask
import com.chess.cryptobot.task.CoinImageTask

class BalanceHolder(fragment: Fragment) : ContextHolder(fragment) {
    private var hasKeys: Boolean = false
    private var bittrexStatuses: Map<String, Boolean> = HashMap()
    private var binanceStatuses: Map<String, Boolean> = HashMap()
    private var poloniexStatuses: Map<String, Boolean> = HashMap()
    private var iconUrls: Map<String, String?> = HashMap()
    var availableCoins = ArrayList<String>()

    init {
        super.init()
        hasKeys = checkIfHasKeys()
    }

    private fun checkIfHasKeys(): Boolean {
        for (market in MarketFactory.getInstance(context).getMarkets()) {
            if (market.keysIsEmpty()) {
                return false
            }
        }
        return true
    }

    public override fun initPrefs(context: Context): Preferences {
        prefs = BalancePreferences(context)
        return prefs
    }

    public override fun initViewItems(itemNamesSet: HashSet<String>?):MutableList<ViewItem> {
        itemNamesSet?.forEach { coinName -> addItemToList(Balance(coinName)) }
        return viewItems
    }

    fun add(coinName: String) {
        val balance = Balance(coinName)
        if (!viewItems.contains(balance))
            add(Balance(coinName))
    }

    override fun add(viewItem: ViewItem) {
        super.add(viewItem)
        updateItem(viewItem)
    }

    public override fun updateItem(item: ViewItem) {
        val balance = item as Balance
        balance.setStatuses(binanceStatuses[balance.name]?: true,
                bittrexStatuses[balance.name]?: true,
                poloniexStatuses[balance.name]?: true)
        balance.coinUrl = iconUrls[balance.name]
        updateImage(balance)
        if (hasKeys) updateAmount(balance)
    }

    private fun updateImage(balance: Balance) {
        val task = CoinImageTask(this)
        task.doInBackground(balance)
    }

    private fun updateAmount(balance: Balance) {
        val task = BalanceUpdateTask(this)
        task.doInBackground(balance)
    }

    @Throws(ItemNotFoundException::class)
    fun getBalanceByPosition(position: Int): Balance {
        val balanceActivity = mainFragment
        val coinName = balanceActivity.itemNameByPosition(position)
        return this.getItemByName(coinName) as Balance
    }

    fun setCurrencyStatus(bittrexStatuses: Map<String, Boolean>, binanceStatuses: Map<String, Boolean>, poloniexStatuses: Map<String, Boolean>) {
        this.bittrexStatuses = bittrexStatuses
        this.binanceStatuses = binanceStatuses
        this.poloniexStatuses = poloniexStatuses
    }

    fun setIconUrls(urls: Map<String, String?>) {
        this.iconUrls = urls
    }
}
